Lib: SMF: Add return code to signal event propagation

glenn-andrews opened the pull request:
See Discussion https://github.com/zephyrproject-rtos/zephyr/discussions/83659 for information about the purpose of this change.

Modifies run actions of state machines to return a value indicating if the event was handled by the run action or should be propagated up to the parent run action.

smf_set_handled() has been removed and replaced by this return value. smf_set_state() will not propagate events regardless of the return value as the transition is considered to have occurred.

Documentation, tests and samples have been updated, as has the hawkBit and USB-C subsystems.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -1606,60 +1608,22 @@ static void s_terminate(void *o)
-		s_terminate,
-		NULL,
-		NULL,
-		NULL,
-		NULL),
+	[S_HAWKBIT_START] = SMF_CREATE_STATE(s_start, NULL, s_end, NULL, NULL),
+	[S_HAWKBIT_HTTP] = SMF_CREATE_STATE(s_http_start, NULL, s_http_end,
+					    &hawkbit_states[S_HAWKBIT_START], NULL),
+	[S_HAWKBIT_PROBE] =
+		SMF_CREATE_STATE(NULL, s_probe, NULL, &hawkbit_states[S_HAWKBIT_HTTP], NULL),
+	[S_HAWKBIT_CONFIG_DEVICE] = SMF_CREATE_STATE(NULL, s_config_device, NULL,
+						     &hawkbit_states[S_HAWKBIT_HTTP], NULL),
+	[S_HAWKBIT_CANCEL] =
+		SMF_CREATE_STATE(NULL, s_cancel, NULL, &hawkbit_states[S_HAWKBIT_HTTP], NULL),
+	[S_HAWKBIT_PROBE_DEPLOYMENT_BASE] = SMF_CREATE_STATE(NULL, s_probe_deployment_base, NULL,
+							     &hawkbit_states[S_HAWKBIT_HTTP], NULL),
+	[S_HAWKBIT_REPORT] =
+		SMF_CREATE_STATE(NULL, s_report, NULL, &hawkbit_states[S_HAWKBIT_HTTP], NULL),
+	[S_HAWKBIT_DOWNLOAD] = SMF_CREATE_STATE(s_download_start, s_download, s_download_end,
+						&hawkbit_states[S_HAWKBIT_HTTP], NULL),
+	[S_HAWKBIT_TERMINATE] = SMF_CREATE_STATE(s_terminate, NULL, NULL, NULL, NULL),
```
Please leave this formated like it was, that was looking better

glenn-andrews replied:
I caught a few it formatted. I'll wrap them in `/* clang-format off */`

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -899,36 +896,35 @@ static void response_json_cb(struct http_response *rsp, enum http_final_call fin
+			ret = json_obj_parse(hb_context->response_data,
+					     hb_context->dl.downloaded_size, json_ctl_res_descr,
+					     ARRAY_SIZE(json_ctl_res_descr),
+					     &hb_context->results.base);
```
Please move the reformating (not just this) into its own commit

maass-hamburg replied:
One commit per subsystem

glenn-andrews replied:
Not sure how to move just the formatting without reapplying my changes manually.

maass-hamburg replied:
Maybe you have to do it manually. Oh and please put the changes in the other subsystems, that are because of the change in the smf in the same commit to allow `git bisect` and `git cherry-pick` at any point.

glenn-andrews replied:
> One commit per subsystem

>  Oh and please put the changes in the other subsystems, that are because of the change in the smf in the same commit to allow git bisect and git cherry-pick at any point.

Which do you want?

maass-hamburg replied:
leave the `apply clang-format` like they are and combine the others. It should be 3 commits then.

glenn-andrews committed:
Lib: SMF: Add return code to signal event propagation

See Discussion https://github.com/zephyrproject-rtos/zephyr/discussions/83659
for information about the purpose of this change.

Modifies run actions of hierarchical state machines
to return a value indicating if the event was handled
by the run action or should be propagated up to the
parent run action. Flat state machines are not affected,
and their run action returns void.

smf_set_handled() has been removed and replaced by
this return value. smf_set_state() will not propagate
events regardless of the return value as the transition
is considered to have occurred.

Documentation, tests, samples, has been updated.
USB-C and hawkBit use SMF and have been updated to use
the new return codes.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews committed:
Lib: USB-C: Apply clang-format

Apply clang-format to USB-C in preparation for code changes

No code is changed, only formatting.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews committed:
Lib: hawkBit: apply clang-format

Apply clang-format in preparation for code changes.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -1293,13 +1293,14 @@ static void s_probe(void *o)
 		hawkbit_event_raise(HAWKBIT_EVENT_NO_UPDATE);
 		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
 	}
+	return SMF_EVENT_PROPAGATE;
```
why `SMF_EVENT_PROPAGATE` and not `SMF_EVENT_HANDLED`, like in all other cases?

maass-hamburg commented on include/zephyr/smf.h:
```diff
@@ -32,14 +32,12 @@
+	 .run = _run,                                                                              \
+	 .exit = _exit,                                                                            \
+	 IF_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT, (.parent = _parent,))                             \
+			    IF_ENABLED(CONFIG_SMF_INITIAL_TRANSITION, (.initial = _initial,)) }
```
this line should not be intended more than the one above, clang format should more be seen as a suggestion and not a a mandatory change.

glenn-andrews replied:
If everyone used clang-format I'd not have needed the first two commits, and every future commit would be clean of formatting changes.

I'd rather machine-assisted uniformity, than others making it 'pretty' and me breaking it every time I want to ensure my code meets compliance.

maass-hamburg commented on include/zephyr/smf.h:
```diff
@@ -32,14 +32,12 @@
+	{.entry = _entry,                                                                          \
+	 .run = _run,                                                                              \
+	 .exit = _exit,                                                                            \
+	 IF_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT, (.parent = _parent,))                             \
+			    IF_ENABLED(CONFIG_SMF_INITIAL_TRANSITION, (.initial = _initial,)) }
```
```suggestion
	 IF_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT, (.parent = _parent,))                             \
	 IF_ENABLED(CONFIG_SMF_INITIAL_TRANSITION, (.initial = _initial,)) }
```

maass-hamburg replied:
yes

maass-hamburg commented on doc/services/smf/index.rst:
```diff
@@ -114,9 +119,15 @@ smf_run_state if it returns a non-zero value.
+:c:enum:`SMF_EVENT_PROPAGATE` is returned.
+
+If :kconfig:option:`CONFIG_SMF_ANCESTOR_SUPPORT` is not enabled, the return
+value from run actions is void as there are no parents to call.
```
The the type of the return value should not depend on that Kconfig. Enabling `CONFIG_SMF_ANCESTOR_SUPPORT` should not break state machines, that don't need that. It could be that you use different sate mashines, where one needs it and the other one does not. both should work at the same time.

glenn-andrews replied:
Drat, you're right. I was hoping not to break people's flat state machines, but of course there's no way of determining mixed flat and hierarchical in the same application.

glenn-andrews replied:
OK, fixed.

glenn-andrews replied:
Are we good now?

keith-zephyr commented:
Can you move the clang-format commits into a separate PR?  The SMF API change needs approval by the Architecture WG, but the clang-format changes don't need to be blocked on that.

keith-zephyr commented:
>  The SMF API change needs approval by the Architecture WG, but the clang-format changes don't need to be blocked on that.

Correcting myself here - unstable APIs do not need approval to modify.  However, I think a raising this issue with the Architecture WG is still a good idea to provide more visibility to the proposed change.

keith-zephyr commented:
> > Can you move the clang-format commits into a separate PR?
> 
> Do they need to be made at all? Seems the use of clang-format is a user preference and I don't think people would like me reformatting their modules for no good reason. The only reason I'm leaving them in is because I'd likely make more mistakes un-formatting them than not.

clang-format is not currently mandated by the project as a whole, but I will always provide a +1 on such changes. :smile:

glenn-andrews commented:
> Correcting myself here - unstable APIs do not need approval to modify. However, I think a raising this issue with the Architecture WG is still a good idea to provide more visibility to the proposed change.

I'm good with that. I can come along next Tuesday if you want?

henrikbrixandersen commented:
> I'm good with that. I can come along next Tuesday if you want?

Sounds good. /CC @carlescufi

kartben commented:
As far as I know SMF is experimental so a migration guide entry might be enough
It's great to see it promoted to unstable though! https://github.com/zephyrproject-rtos/zephyr/pull/83644

glenn-andrews commented:
@keith-zephyr pointed out we have other projects depending on it so it should have some thought to it: https://github.com/zephyrproject-rtos/zephyr/discussions/83659#discussioncomment-11856584

carlescufi reviewed (changes requested):
Please add an entry in the migration document

carlescufi commented:
Architecture WG:

- @glenn-andrews explains the change, which essentially simplifies the handling of substates by not having to explicitly call `smf_set_handled()`
- @henrikbrixandersen agrees with the change
- @glenn-andrews suggest to make this change (and another) before transitioning the API from experimental to unstable

glenn-andrews commented:
How are things looking now?

keith-zephyr commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -524,6 +524,15 @@ LoRa
+State Machine Framework
+=======================
+
+* State run actions now return an :c:enum:`smf_state_result` value instead of void.
+  :c:func:`smf_set_handled` has been removed, and the return code determines if the
+  event is propagated to parent run actions or has been handled in this run action.
```
I suggest flipping the order here - mention that `smf_set_handled` has been removed first and then describe the changes to state run actions.
It would also be helpful to provide instructions on how to transition an existing hierarchical state machine - something like "run states called smf_set_handled() should return SMF_EVENT_HANDLED and otherwise should return SMF_EVENT_PROPAGATE."

keith-zephyr commented on doc/services/smf/index.rst:
```diff
@@ -39,6 +40,16 @@ By default, a state can have no ancestor states, resulting in a flat state
 :kconfig:option:`CONFIG_SMF_ANCESTOR_SUPPORT` option must be enabled.
 
+The return value of the run action, :c:enum:`smf_state_result` determines if the
+state machine propagate the event to parent run actions
```
```suggestion
state machine propagates the event to parent run actions
```

keith-zephyr commented on doc/services/smf/index.rst:
```diff
@@ -39,6 +40,16 @@ By default, a state can have no ancestor states, resulting in a flat state
 
+The return value of the run action, :c:enum:`smf_state_result` determines if the
+state machine propagate the event to parent run actions
+(:c:enum:`SMF_EVENT_PROPAGATE`) or of the event was handled by the run action
```
```suggestion
(:c:enum:`SMF_EVENT_PROPAGATE`) or if the event was handled by the run action
```

glenn-andrews commented:
> Please add an entry in the migration document

Done

glenn-andrews commented:
Now Zephyr 4.1 is out, can we revisit this change?

fabiobaltieri commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -728,6 +728,17 @@ Secure Storage
   and settings, which means the NVS settings backend would get used by default if ZMS wasn't
   enabled. (:github:`86181`)
 
+State Machine Framework
+=======================
+
+* :c:func:`smf_set_handled` has been removed.
+* State run actions now return an :c:enum:`smf_state_result` value instead of void. and the return
+  code determines if the event is propagated to parent run actions or has been handled. A run action
+  that handles the event completely should return :c:enum:`SMF_EVENT_HANDLED`, and run actions that
+  propagate handling to parent states should return :c:enum:`SMF_EVENT_PROPAGATE`.
+* Flat state machines ignore the return value; returning :c:enum:`SMF_EVENT_HANDLED`
+  would be the most technically accurate response.
+
```
this needs to go on the 4.2 file

glenn-andrews replied:
Done.

glenn-andrews commented:
@carlescufi @maass-hamburg Are the requested changes complete to your satisfaction?

fabiobaltieri commented:
@sambhurst can you take a look?

maass-hamburg commented:
@glenn-andrews rebase needed

glenn-andrews commented:
> @glenn-andrews rebase needed

I think I screwed up the rebase. Help!

kartben commented:
> > @glenn-andrews rebase needed
> 
> I think I screwed up the rebase. Help!

should be good now :)

maass-hamburg commented on samples/subsys/smf/smf_calculator/src/smf_calculator_thread.c:
```diff
@@ -352,7 +359,7 @@ static void zero_1_run(void *obj)
 	case DIGIT_0:
 		/* We ignore leading zeroes */
-		smf_set_handled(&s->ctx);
+		return SMF_EVENT_HANDLED;
 		break;
```
break after return

maass-hamburg commented on samples/subsys/smf/smf_calculator/src/smf_calculator_thread.c:
```diff
@@ -508,7 +520,7 @@ static void zero_2_run(void *obj)
 	case DIGIT_0:
 		/* We ignore leading zeroes */
-		smf_set_handled(&s->ctx);
+		return SMF_EVENT_HANDLED;
 		break;
```
here too

maass-hamburg commented on samples/subsys/smf/smf_calculator/src/smf_calculator_thread.c:
```diff
@@ -534,7 +547,7 @@ static void int_2_run(void *obj)
 	case DIGIT_1_9:
 		insert(&s->operand_2, s->event.operand);
-		smf_set_handled(&s->ctx);
+		return SMF_EVENT_HANDLED;
 		break;
```
here too

maass-hamburg commented on samples/subsys/smf/smf_calculator/src/smf_calculator_thread.c:
```diff
@@ -556,19 +570,20 @@ static void frac_2_run(void *obj)
 		insert(&s->operand_2, s->event.operand);
-		smf_set_handled(&s->ctx);
+		return SMF_EVENT_HANDLED;
 		break;
 	case DECIMAL_POINT:
 		/* Ignore further decimal points */
-		smf_set_handled(&s->ctx);
+		return SMF_EVENT_HANDLED;
 		break;
 	default:
```
here too

maass-hamburg commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -858,21 +854,22 @@ static void pe_drs_send_swap_run(void *obj)
 		pe_set_ready_state(dev);
-		return;
+		return SMF_EVENT_PROPAGATE;
 	} else if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_MSG_DISCARDED)) {
 		/*
 		 * Inform Device Policy Manager that the message
 		 * was discarded
 		 */
 		policy_notify(dev, MSG_DISCARDED);
 		pe_set_ready_state(dev);
-		return;
+		return SMF_EVENT_PROPAGATE;
```
both of these returns seem unneeded, as the same return is directly after the ifs end

maass-hamburg commented on subsys/usb/usb_c/usbc_pe_snk_states.c:
```diff
@@ -317,8 +320,9 @@ void pe_snk_select_capability_run(void *obj)
 			pe_send_soft_reset(dev, prl_rx->emsg.type);
 		}
-		return;
+		return SMF_EVENT_PROPAGATE;
 	}
+	return SMF_EVENT_PROPAGATE;
```
first return seems unnneded, only is in a if

maass-hamburg commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -662,8 +662,9 @@ static void prl_tx_wait_for_message_request_run(void *obj)
 			prl_tx_set_state(dev, PRL_TX_WAIT_FOR_PHY_RESPONSE);
 		}
-		return;
+		return SMF_EVENT_PROPAGATE;
 	}
+	return SMF_EVENT_PROPAGATE;
```
same here

maass-hamburg commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -748,8 +749,9 @@ static void prl_tx_wait_for_phy_response_run(void *obj)
-		return;
+		return SMF_EVENT_PROPAGATE;
 	}
+	return SMF_EVENT_PROPAGATE;
```
the 2 returns before seem redundant

glenn-andrews replied:
Same response.

glenn-andrews commented:
@maass-hamburg You OK with the changes?

maass-hamburg commented on tests/lib/smf/src/test_lib_hierarchical_5_ancestor_smf.c:
```diff
@@ -35,34 +35,34 @@
-#define TEST_VALUE_NUM              24
+#define SMF_RUN 3
+
+#define P05_ENTRY_BIT (1 << 0)
+#define P04_ENTRY_BIT (1 << 1)
+#define P03_ENTRY_BIT (1 << 2)
+#define P02_ENTRY_BIT (1 << 3)
+#define P01_ENTRY_BIT (1 << 4)
+#define A_ENTRY_BIT   (1 << 5)
+#define A_RUN_BIT     (1 << 6)
+#define A_EXIT_BIT    (1 << 7)
+#define B_ENTRY_BIT   (1 << 8)
+#define B_RUN_BIT     (1 << 9)
+#define P01_RUN_BIT   (1 << 10)
+#define P02_RUN_BIT   (1 << 11)
+#define P03_RUN_BIT   (1 << 12)
+#define P04_RUN_BIT   (1 << 13)
+#define P05_RUN_BIT   (1 << 14)
+#define B_EXIT_BIT    (1 << 15)
+#define P01_EXIT_BIT  (1 << 16)
+#define P02_EXIT_BIT  (1 << 17)
+#define P03_EXIT_BIT  (1 << 18)
+#define P04_EXIT_BIT  (1 << 19)
+#define P05_EXIT_BIT  (1 << 20)
+#define C_ENTRY_BIT   (1 << 21)
+#define C_RUN_BIT     (1 << 22)
+#define C_EXIT_BIT    (1 << 23)
```
if you are changing it already please use `BIT()`

maass-hamburg commented on tests/lib/smf/src/test_lib_hierarchical_smf.c:
```diff
@@ -41,24 +41,24 @@
-#define STATE_B_ENTRY_BIT       (1 << 5)
-#define STATE_B_RUN_BIT         (1 << 6)
-#define STATE_B_EXIT_BIT        (1 << 7)
-#define PARENT_AB_EXIT_BIT      (1 << 8)
+#define STATE_B_ENTRY_BIT  (1 << 5)
+#define STATE_B_RUN_BIT    (1 << 6)
+#define STATE_B_EXIT_BIT   (1 << 7)
+#define PARENT_AB_EXIT_BIT (1 << 8)
 
-#define PARENT_C_ENTRY_BIT      (1 << 9)
-#define STATE_C_ENTRY_BIT       (1 << 10)
-#define STATE_C_RUN_BIT         (1 << 11)
-#define STATE_C_EXIT_BIT        (1 << 12)
-#define PARENT_C_EXIT_BIT       (1 << 13)
+#define PARENT_C_ENTRY_BIT (1 << 9)
+#define STATE_C_ENTRY_BIT  (1 << 10)
+#define STATE_C_RUN_BIT    (1 << 11)
+#define STATE_C_EXIT_BIT   (1 << 12)
+#define PARENT_C_EXIT_BIT  (1 << 13)
```
here too

glenn-andrews commented:
> @glenn-andrews smf ci tests are failing

Yep, and on Windows the `qemu_cortex_r5/zynqmp_rpu` fails to build entirely so I have no idea what's going on.

Can someone run ` west twister -p qemu_cortex_r5/zynqmp_rpu -s sample.smf.hsm_psicc2 --no-detailed-test-id` under Linux and see what the failure is?

The tests are passing for other qemu targets.

```
(.venv) PS C:\Users\glenn\zephyrproject_latest\zephyr\tests\lib\smf> west twister -p qemu_cortex_r5/zynqmp_rpu -s sample.smf.hsm_psicc2 --no-detailed-test-id --short-build-path
INFO    - Using Ninja..
INFO    - Zephyr version: v4.1.0-5262-g6df4056ba214
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report C:\Users\glenn\zephyrproject_latest\zephyr\tests\lib\smf\twister-out\testplan.json
Junction created for C:\Users\glenn\zephyrproject_latest\zephyr\tests\lib\smf\twister-out\twister_links\test_0 <<===>> C:\Users\glenn\zephyrproject_latest\zephyr\tests\lib\smf\twister-out\qemu_cortex_r5_zynqmp_rpu\zephyr\samples\subsys\smf\hsm_psicc2\sample.smf.hsm_psicc2
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
ERROR   - General exception: 'NoneType' object cannot be interpreted as an integer
Traceback (most recent call last):
  File "C:\Users\glenn\zephyrproject_latest\zephyr\scripts\pylib\twister\twisterlib\runner.py", line 2023, in pipeline_mgr
    pb.process(pipeline, done_queue, task, lock, results)
  File "C:\Users\glenn\zephyrproject_latest\zephyr\scripts\pylib\twister\twisterlib\runner.py", line 1129, in process
    self.run()
  File "C:\Users\glenn\zephyrproject_latest\zephyr\scripts\pylib\twister\twisterlib\runner.py", line 1790, in run
    instance.handler.handle(harness)
  File "C:\Users\glenn\zephyrproject_latest\zephyr\scripts\pylib\twister\twisterlib\handlers.py", line 1427, in handle
    os.close(self.pipe_handle)
TypeError: 'NoneType' object cannot be interpreted as an integer
```

glenn-andrews commented:
@brandon-exact Thank you so much!

glenn-andrews commented:
> @glenn-andrews smf ci tests are failing

@maass-hamburg CI should be passing now, thanks to @brandon-exact

keith-zephyr commented:
> @sambhurst, @keith-zephyr: Please take another look.

I'm reviewing now.  I have also pulled in this change locally to test with our downstream repo as we have a few uses of SMF in our application.
